hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AddMigrationListenerMessageTask.java

```diff
@@ -0,0 +1,153 @@
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.partition.IPartitionService;
+import com.hazelcast.partition.MigrationListener;
+import com.hazelcast.partition.MigrationState;
+import com.hazelcast.internal.partition.MigrationStateImpl;
+import com.hazelcast.partition.ReplicaMigrationEvent;
+
+import java.security.Permission;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import static com.hazelcast.internal.partition.InternalPartitionService.MIGRATION_EVENT_TOPIC;
+import static com.hazelcast.internal.partition.InternalPartitionService.MIGRATION_EVENT_TOPIC_ORDER_KEY;
+import static com.hazelcast.internal.partition.MigrationEventHandler.MIGRATION_FINISHED;
+import static com.hazelcast.internal.partition.MigrationEventHandler.MIGRATION_STARTED;
+import static com.hazelcast.spi.impl.InternalCompletableFuture.newCompletedFuture;
+
+public class AddMigrationListenerMessageTask
+        extends AbstractAddListenerMessageTask<Boolean> {
+
+    public AddMigrationListenerMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected CompletableFuture<UUID> processInternal() {
+        IPartitionService partitionService = getService(getServiceName());
+
+        MigrationListener listener = createMigrationListener();
+
+        if (parameters) {
+            return newCompletedFuture(partitionService.addLocalMigrationListener(listener));
+        }
+
+        return partitionService.addMigrationListenerAsync(listener);
+    }
+
+    private MigrationListener createMigrationListener() {
+        return new MigrationListener() {
+
+                @Override
+                public void migrationStarted(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, MIGRATION_STARTED));
+                }
+
+                @Override
+                public void migrationFinished(MigrationState state) {
+                    sendIfAlive(encodeMigrationEvent(state, MIGRATION_FINISHED));
+                }
+
+                @Override
+                public void replicaMigrationCompleted(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+
+                @Override
+                public void replicaMigrationFailed(ReplicaMigrationEvent event) {
+                    sendIfAlive(encodeReplicaMigrationEvent(event));
+                }
+            };
+    }
+
+    private void sendIfAlive(ClientMessage eventMessage) {
+        if (endpoint.isAlive()) {
+            sendClientMessage(MIGRATION_EVENT_TOPIC_ORDER_KEY, eventMessage);
+        }
+    }
+
+    private ClientMessage encodeReplicaMigrationEvent(ReplicaMigrationEvent event) {
+        return ClientAddMigrationListenerCodec.encodeReplicaMigrationEvent(
+                new MigrationStateImpl(
```

Comment:
No need to create new `MigrationStateImpl` object. We can pass event.getMigrationState to `encode..` method